pyface/ui/qt4/data_view/data_view_item_model.py

```diff
@@ -0,0 +1,239 @@
+                'values_changed',
+                remove=True,
+            )
+
+            self.beginResetModel()
+            self._model = model
+            self.endResetModel()
+        else:
+            # model is being initialized
+            self._model = model
+
+        # hook up trait listeners
+        self._model.observe(
+            self.on_structure_changed,
+            'structure_changed',
+        )
+        self._model.observe(
+            self.on_values_changed,
+            'values_changed',
+        )
+
+    # model event listeners
+
+    def on_structure_changed(self, event):
+        self.beginResetModel()
+        self.endResetModel()
+
+    def on_values_changed(self, event):
+        top, left, bottom, right = event.new
+        if top == () and bottom == ():
+            # this is a column header change
+            self.headerDataChanged.emit(left[0], right[0])
+        elif left == () and right == ():
+            # this is a row header change
+            # XXX this is currently not supported and not needed
+            pass
+        else:
+            for i, (top_row, bottom_row) in enumerate(zip(top, bottom)):
+                if top_row != bottom_row:
+                    break
+            top = top[:i+1]
+            bottom = bottom[:i+1]
+
+            top_left = self._to_model_index(top, left)
+            bottom_right = self._to_model_index(bottom, right)
+            self.dataChanged.emit(top_left, bottom_right)
+
+    # Structure methods
+
+    def parent(self, index):
+        if not index.isValid():
+            return QModelIndex()
+        parent = index.internalPointer()
+        if parent == Root:
+            return QModelIndex()
+
+        grandparent, row = self.model.index_manager.get_parent_and_row(parent)
+        return self.createIndex(row, 0, grandparent)
+
+    def index(self, row, column, parent):
+        if parent.isValid():
+            parent_index = self.model.index_manager.create_index(
+                parent.internalPointer(),
+                parent.row(),
+            )
+        else:
+            parent_index = Root
+        index = self.createIndex(row, column, parent_index)
+        return index
+
+    def rowCount(self, index):
+        row_index = self._to_row_index(index)
+        try:
+            if self.model.can_have_children(row_index):
+                return self.model.get_row_count(row_index)
+        except Exception:
+            logger.exception("Error in rowCount")
+
+        return 0
+
+    def columnCount(self, index):
+        row_index = self._to_row_index(index)
+        try:
+            # the number of columns is constant; leaf rows return 0
+            if self.model.can_have_children(row_index):
+                return self.model.get_column_count() + 1
+            else:
+                return 0
+        except Exception:
+            logger.exception("Error in columnCount")
+
+    # Data methods
+
+    def flags(self, index):
```

Comment:
Curiously, that is not documented anywhere in the Qt docs.

Reply:
On documentation: perhaps because the flags themselves are documented? e.g. One may think that the `Qt.NoItemFlags` is obviously appropriate for an invalid index and hence no need to document that again on `flags`.
See https://bugreports.qt.io/browse/QTBUG-18654

I see the returned value is now 0 (==NoItemFlags) if the index is invalid. Resolving.